File: algorithm/cpp/prob1501-2000/prob1799_hard_N次操作后的最大分数和.cpp
class Solution {
public:
    int maxScore(vector<int>& nums) {
        int n = nums.size();
        dp = vector<int>(1 << n, -1);
        return solve(0, nums, 1);
    }

private:
    vector<int> dp;
    int solve(int state, const vector<int>& nums, int r)
    {
        if(dp[state] != -1)
            return dp[state];
        int n = nums.size();
        int ans = 0;
        for(int i = 0; i < n; ++i)
        {
            if(state >> i & 1)
                continue;
            for(int j = i + 1; j < n; ++j)
            {
                if(state >> j & 1)
                    continue;
                ans = max(ans, r * (gcd<int>(nums[i], nums[j])) + solve(state | (1 << i) | (1 << j), nums, r + 1));
            }
        }
        dp[state] = ans;
        return ans;
    }
};

File: algorithm/python/prob1501-2000/prob1799_hard_N次操作后的最大分数和.py
class Solution:
    def maxScore(self, nums: List[int]) -> int:
        self.nums = nums
        self.n = len(nums)
        return self.solve(0, 1)

    @lru_cache(int(1e7))
    def solve(self, state: int, r: int) -> int:
        ans = 0
        for i in range(self.n):
            if state >> i & 1:
                continue
            for j in range(i + 1, self.n):
                if state >> j & 1:
                    continue
                ans = max(ans, r * gcd(self.nums[i], self.nums[j]) + self.solve(state | (1 << i) | (1 << j), r + 1))
        return ans
